File: app/static/js/stock_management.js
$(document).ready(function() {
    $('#usernameDisplay').text(sessionStorage.getItem('username', 'Farmer'));

    $(document).on('click', '#logout', function() {
        sessionStorage.clear();
        localStorage.clear();
        window.location.href = '/';
    });

    function loadItems() {
        $.ajax({
            url: "/api/stock_management/list_items",
            method: "GET",
            beforeSend: function(xhr) {
                var farmerId = sessionStorage.getItem('farmerId');
                xhr.setRequestHeader('farmer-id', farmerId);
            },
            success: function(data) {
                $('#itemList').empty();
                data.items.forEach(item => {
                    $('#itemList').append(
                        `<tr>
                            <td>${item.name}</td>
                            <td>${item.quantity}</td>
                            <td>${item.validity}</td>
                            <td>₪${item.price}</td>
                            <td>${item.category}</td>
                            <td>
                                <span class="icon" onclick="deleteItem(${item.id})">&#128465;</span>
                                <span class="icon" onclick="editItem(${item.id}, '${item.name}', ${item.price}, ${item.quantity}, '${item.validity}')">&#9998;</span>
                            </td>
                        </tr>`
                    );
                });
            }
        });
    }

    function loadProductNames() {
        $.ajax({
            url: "/api/stock_management/list_product_names",
            method: "GET",
            success: function(data) {
                $('#productNames').empty();
                data.productNames.forEach(name => {
                    $('#productNames').append(
                        `<option value="${name}">${name}</option>`
                    );
                });
            }
        });
    }

    $('#itemForm').submit(function(e) {
        e.preventDefault();
        var id = $('#itemId').val();
        var url = "/api/stock_management/add_item";
        var method = "POST";
        if (id) {
            url = `/api/stock_management/update_item/${id}`;
            method = "PUT";
        }
        var itemData = {
            name: $('#itemName').val(),
            price: $('#itemPrice').val(),
            quantity: $('#itemQuantity').val(),
            validity: $('#itemValidity').val(),
        };
        $.ajax({
            url: url,
            method: method,
            contentType: "application/json",
            data: JSON.stringify(itemData),
            beforeSend: function(xhr) {
                var farmerId = sessionStorage.getItem('farmerId');
                xhr.setRequestHeader('farmer-id', farmerId);
            },
            success: function() {
                loadItems();
                $('#itemForm').trigger("reset");
                $('#itemModal').modal('hide');
                $('#itemId').val('');
            }
        });
    });

    window.deleteItem = function(id) {
        $.ajax({
            url: `/api/stock_management/delete_item/${id}`,
            method: "DELETE",
            beforeSend: function(xhr) {
                var farmerId = sessionStorage.getItem('farmerId');
                xhr.setRequestHeader('farmer-id', farmerId);
            },
            success: function() {
                loadItems();
            }
        });
    };

    window.editItem = function(id, name, price, quantity, validity) {
        $('#itemId').val(id);
        $('#itemName').val(name);
        $('#itemPrice').val(price);
        $('#itemQuantity').val(quantity);
        $('#itemValidity').val(validity);
        $('#itemModalLabel').text('Edit Item');
        $('#itemModal').modal('show');
    };

    $('#itemModal').on('shown.bs.modal', function () {
        loadProductNames();
    });

    $('#itemModal').on('hidden.bs.modal', function () {
        $('#itemModalLabel').text('Add New Item');
        $('#itemForm').trigger("reset");
        $('#itemId').val('');
    });

    loadItems();
});

File: app/stock_management.py
from sqlalchemy.orm import scoped_session, sessionmaker
from .globals import Globals
from .models import Product


class StockManager:
    def __init__(self, farmer_id):
        self.farmer_id = farmer_id
        self.db_session = scoped_session(sessionmaker(bind=Globals.db.engine))

    def add_item(self, item_data):
        new_item = Product(
            name=item_data['name'],
            price=item_data['price'],
            quantity=item_data['quantity'],
            validity=item_data['validity'],
            farmer_id=self.farmer_id
        )
        self.db_session.add(new_item)
        self.db_session.commit()
        return new_item

    def update_item(self, item_id, update_data):
        item = self.db_session.query(Product).filter_by(id=item_id).first()
        if item:
            for key, value in update_data.items():
                setattr(item, key, value)
            self.db_session.commit()
            return item
        return None

    def delete_item(self, item_id):
        item = self.db_session.query(Product).filter_by(id=item_id).first()
        if item:
            self.db_session.delete(item)
            self.db_session.commit()
            return True
        return False

    def list_items(self):
        query = self.db_session.query(Product).filter_by(farmer_id=self.farmer_id)
        return query.all()
